test/framework/src/main/java/org/elasticsearch/test/disruption/DisruptableMockTransport.java

```diff
@@ -42,58 +59,79 @@ public DisruptableMockTransport(Logger logger) {
 
-    protected abstract Optional<DisruptableMockTransport> getDisruptedCapturingTransport(DiscoveryNode node, String action);
+    protected abstract Optional<DisruptableMockTransport> getDisruptableMockTransport(TransportAddress address);
 
-    protected abstract void handle(DiscoveryNode sender, DiscoveryNode destination, String action, Runnable doDelivery);
+    protected abstract void schedule(Runnable runnable);
 
-    protected final void sendFromTo(DiscoveryNode sender, DiscoveryNode destination, String action, Runnable doDelivery) {
-        handle(sender, destination, action, new Runnable() {
-            @Override
-            public void run() {
-                if (getDisruptedCapturingTransport(destination, action).isPresent()) {
-                    doDelivery.run();
-                } else {
-                    logger.trace("unknown destination in {}", this);
-                }
-            }
+    protected final void schedule(String action, Runnable runnable) {
+        // handshake needs to run inline as the caller blockingly waits on the result
+        if (action.equals(HANDSHAKE_ACTION_NAME)) {
+            runnable.run();
+        } else {
+            schedule(runnable);
+        }
+    }
 
-            @Override
-            public String toString() {
-                return doDelivery.toString();
-            }
-        });
+    @Override
+    public TransportService createTransportService(Settings settings, ThreadPool threadPool, TransportInterceptor interceptor,
+                                                   Function<BoundTransportAddress, DiscoveryNode> localNodeFactory,
+                                                   @Nullable ClusterSettings clusterSettings, Set<String> taskHeaders) {
+        return new TransportService(settings, this, threadPool, interceptor, localNodeFactory, clusterSettings, taskHeaders);
     }
 
     @Override
-    protected void onSendRequest(long requestId, String action, TransportRequest request, DiscoveryNode destination) {
+    public Releasable openConnection(DiscoveryNode node, ConnectionProfile profile, ActionListener<Connection> listener) {
+        final Optional<DisruptableMockTransport> matchingTransport = getDisruptableMockTransport(node.getAddress());
+        if (matchingTransport.isPresent()) {
+            listener.onResponse(new CloseableConnection() {
+                @Override
+                public DiscoveryNode getNode() {
+                    return node;
+                }
+
+                @Override
+                public void sendRequest(long requestId, String action, TransportRequest request, TransportRequestOptions options)
+                    throws TransportException {
+                    onSendRequest(requestId, action, request, matchingTransport.get());
+                }
+            });
+            return () -> {};
+        } else {
+            throw new ConnectTransportException(node, "node " + node + " does not exist");
```

Comment:
I think this is OK for now but in future (hoho) we will want this to be async and/or to timeout on an unknown node.

Reply:
this depends on the connection manager becoming async. Right now there's a `Future.get()` waiting for us behind this call.